### PR TITLE
sql/schemachanger/scplan/internal/rules: rework dep rules

### PR DESCRIPTION
**Pull request by ajwerner**
The dep rules helpers were brittle. Instead, make it easier to construct rel
queries directly.

Release note: None

### CODE REVIEW

File: pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go [hunk 1]
```diff
@@ -358,37 +383,45 @@ func init() {
...
-	}).register()
+		"partial-predicate", "index",
+		func(from, fromTarget, fromNode, to, toTarget, toNode rel.Var) rel.Clauses {
+			return rel.Clauses{
+				from.Type((*scpb.SecondaryIndexPartial)(nil)),
+				to.Type((*scpb.SecondaryIndex)(nil)),
+				joinOnIndexID(from, to),
+				targetStatusEq(fromTarget, toTarget, scpb.ToAbsent),
+				currentStatusEq(fromNode, toNode, scpb.Status_ABSENT),
+				rel.Filter("secondaryIndexPartialIsNotBeingDropped", from)(func(
```
Comment (postamar) [inline pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:417]: Kebab-case all the rel things?

File: pkg/sql/schemachanger/scplan/internal/rules/helpers.go [hunk 2]
```diff
@@ -32,93 +30,43 @@ func targetNodeVars(el rel.Var) (element, target, node rel.Var) {
...
-	c = append(c, to.AttrEqVar(screl.DescID, "var-to-tell-rel-to-is-an-element"))
+}
 
-	if d.filter != nil {
-		c = append(c, rel.Filter(d.filterLabel, from, to)(d.filter))
-	}
-	registerDepRule(d.ruleName, d.edgeKind, fromNode, toNode, screl.MustQuery(c...))
+func joinOnColumnID(a, b rel.Var) rel.Clause {
+	return rel.And(
+		join(a, b, screl.DescID, "desc-id"),
```
Comment (postamar) [inline pkg/sql/schemachanger/scplan/internal/rules/helpers.go:69]: suggestion: `relation-id`? also applies to `joinOnIndexID` and others

File: pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go [hunk 3]
```diff
@@ -22,319 +22,344 @@ import (
...
-			toNode.AttrEq(screl.CurrentStatus, scpb.Status_BACKFILLED),
-
-			screl.JoinTargetNode(from, fromTarget, fromNode),
-			screl.JoinTargetNode(to, toTarget, toNode),
-		),
+		"temp", "index",
+		func(from, fromTarget, fromNode, to, toTarget, toNode rel.Var) rel.Clauses {
+			return rel.Clauses{
+				from.Type((*scpb.TemporaryIndex)(nil)),
+				to.Type((*scpb.PrimaryIndex)(nil), (*scpb.SecondaryIndex)(nil)),
```
Comment (postamar) [inline pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:201]: I wish there was a way to make these type unions pop out more systematically somehow. It's why I originally newlined each subtype, I found that it helped readability. I think this is fine though.

**Review by postamar — APPROVED** [review]
LGTM, I like it and it's simpler.

File: pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go (hunk 3 above)
Comment (ajwerner) [reply pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:204]: Done.

File: pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go (hunk 1 above)
Comment (ajwerner) [reply pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:420]: I made the filter functions camelCase. I'm starting to dislike the kebabs. If I were typing this language I would not be pleased with kabobs I don't think. Sometimes I think about what an interpreter would look like. I'm going to keep iterating around these parts, happy to make further changes later.

**Comment by ajwerner** [comment]
@postamar I added a new first commit to this which, in retrospect, is an important one. There was a buglet where we'd say that an index can be used even if it matches none of the attributes. This lead to the possibility of writing rules which would not be index accelerated. The hope is to always index accelerate rule evaluation. The immediately caught the bug which lead to the change making logic tests slow. PTAL before I bors it.

**Comment by ajwerner** [comment]
I'm going to bors it.

bors r+

**Pull request merged**